File: tedorilog-poc/engine/save_guard.js
// 保存ガード。
// PoCの最重要ルール「認識値をユーザー確認なしで確定保存しない」をコードで保証する層。
// UIはこの関数を通してしか保存データを作れない。

import { ITEM_KEYS, ITEM_LABELS } from './lexicon.js';
import { STATUS } from './extract.js';

/**
 * @param {object} result extractPayslip の結果
 * @param {Record<string, {value:number|null, confirmed:boolean, edited?:boolean}>} confirmations
 *        画面Bでユーザーが確認・修正した内容
 * @returns {{ok:boolean, blocked:Array<{key:string,label:string,reason:string}>, payload:object|null}}
 */
export function buildSaveDraft(result, confirmations = {}) {
  const blocked = [];
  const entries = {};

  for (const key of ITEM_KEYS) {
    const item = (result && result.items && result.items[key]) || null;
    const input = confirmations[key];
    const suggested = item ? item.value : null;
    const value = !input || input.value === undefined ? null : input.value;
    const confirmed = Boolean(input && input.confirmed);

    if (value === null || value === undefined) {
      if (suggested !== null && !confirmed) {
        // 候補があるのに確認されていない = 画面Bを通っていない。保存させない。
        blocked.push({
          key,
          label: ITEM_LABELS[key],
          reason: '解析候補が未確認です。確認するか、未設定に戻してください',
        });
        continue;
      }
      // 未検出のまま、または明示的に未設定へ戻した場合は空欄で保存できる
      entries[key] = { key, label: ITEM_LABELS[key], value: null, source: 'empty' };
      continue;
    }
    if (!Number.isFinite(value)) {
      blocked.push({ key, label: ITEM_LABELS[key], reason: '数値ではありません' });
      continue;
    }
    if (!confirmed) {
      blocked.push({
        key,
        label: ITEM_LABELS[key],
        reason: item && item.status === STATUS.CONFIRMED_CANDIDATE
          ? '確定候補でも、ユーザー確認なしには保存できません'
          : '要確認の項目です。内容を確認してください',
      });
      continue;
    }
    const edited = input.edited !== undefined ? Boolean(input.edited) : value !== suggested;
    entries[key] = {
      key,
      label: ITEM_LABELS[key],
      value,
      source: edited ? 'user_edited' : 'user_confirmed',
      suggested,
    };
  }

  if (blocked.length) return { ok: false, blocked, payload: null };

  return {
    ok: true,
    blocked: [],
    payload: {
      items: entries,
      route: result ? result.route : null,
      confirmedAt: new Date().toISOString(),
      // 解析メタ情報のみ保持し、明細画像・原文テキストは保存しない
      meta: result
        ? { tokenCount: result.stats.tokenCount, checks: result.checks.map((c) => ({ id: c.id, ok: c.ok })) }
        : null,
    },
  };
}

File: native-ios/TedoriLogInput/Sources/TedoriLogCore/SaveGuard.swift
import Foundation

/// 保存ガード。
/// 「認識値をユーザー確認なしで確定保存しない」をコードで担保する層。
/// UIはこの型を通してしか保存データを作れない。
public enum SaveGuard {

    public struct Confirmation {
        public var value: Int?
        public var confirmed: Bool
        public var edited: Bool?

        public init(value: Int?, confirmed: Bool, edited: Bool? = nil) {
            self.value = value
            self.confirmed = confirmed
            self.edited = edited
        }
    }

    public enum Source: String, Codable {
        case userConfirmed = "user_confirmed"
        case userEdited = "user_edited"
        case empty
    }

    public struct SavedItem: Codable {
        public var key: ItemKey
        public var label: String
        public var value: Int?
        public var source: Source
        public var suggested: Int?
    }

    public struct Payload: Codable {
        public var items: [SavedItem]
        public var route: String
        public var confirmedAt: Date
        /// 明細の画像・原文テキストは保存しない。解析メタ情報のみ。
        public var tokenCount: Int
        public var passedChecks: [String]
    }

    public struct Blocked {
        public var key: ItemKey
        public var label: String
        public var reason: String
    }

    public struct Draft {
        public var ok: Bool
        public var blocked: [Blocked]
        public var payload: Payload?
    }

    public static func buildDraft(result: PayslipResult,
                                  confirmations: [ItemKey: Confirmation]) -> Draft {
        var blocked: [Blocked] = []
        var entries: [SavedItem] = []

        for key in ItemKey.allCases {
            let item = result.items[key]
            let suggested = item?.value
            let input = confirmations[key]
            let value = input?.value
            let confirmed = input?.confirmed ?? false

            guard let value else {
                if suggested != nil && !confirmed {
                    blocked.append(Blocked(key: key, label: key.label,
                                           reason: "解析候補が未確認です。確認するか、未設定に戻してください"))
                } else {
                    entries.append(SavedItem(key: key, label: key.label, value: nil, source: .empty, suggested: suggested))
                }
                continue
            }
            guard confirmed else {
                blocked.append(Blocked(key: key, label: key.label,
                                       reason: item?.status == .confident
                                           ? "確定候補でも、ユーザー確認なしには保存できません"
                                           : "要確認の項目です。内容を確認してください"))
                continue
            }
            let edited = input?.edited ?? (value != suggested)
            entries.append(SavedItem(key: key, label: key.label, value: value,
                                     source: edited ? .userEdited : .userConfirmed, suggested: suggested))
        }

        guard blocked.isEmpty else { return Draft(ok: false, blocked: blocked, payload: nil) }
        let payload = Payload(items: entries, route: result.route, confirmedAt: Date(),
                              tokenCount: result.tokenCount,
                              passedChecks: result.checks.filter(\.ok).map(\.id))
        return Draft(ok: true, blocked: [], payload: payload)
    }
}
